src/main/scala/it/unibo/authsim/library/dsl/cryptography/encrypter/Cipher.scala

```diff
@@ -0,0 +1,112 @@
+   */
+  def decrypt[A, B](encryptedPassword: A, secret:B): String
+  
+/**
+ * Trait for Symmetric encrypter
+ */
+trait SymmetricEncrypter extends Cipher                                            //TODO non so se lasciarla o rimuoverla
+
+/**
+ * Trait for Asymmetric encrypter, it provides additional methods to manage the encryption operation
+ * using asymmetric encryption algorithms
+ */
+trait AsymmetricEncrypter extends Cipher:
+  
+  /**
+   * Method to load existing key from a user directory 
+   * 
+   * @param fileName                name of the file from which key should be loaded
+   * @return                        an istance of KeyPair
+   */
+  def loadKeys(fileName: String):KeyPair
+
+  /**
+   * Method to generate a new key pair and save it on the disk, overriding prexisting information present in the file
+   * 
+   * @param fileName                name of the file in which key pair should be saved
+   * @return                        the key pair generated
+   */
+  def generateKeys(fileName: String): KeyPair
+
+/**
+ * Abstract class for to perform the encryption operation
+ */
+abstract class BasicCipher extends Cipher:
+  
+  /**
+   * Method used to encrypt the password
+   *
+   * @param password              password to be encrypted
+   * @param secret                secret used to encrypt the password
+   * @tparam A                    generic parameter for the password
+   * @tparam B                    generic parameter for the secret
+   *  @return                     a string representing the password encrypted
+   */
+  override def encrypt[A,B](password: A, secret: B): String =
+    crypto(EncryptionMode.Encryption, password, secret)
+
+  /**
+   * Method used to decrypt the password
+   *
+   * @param encryptedPassword      encrypted password to be decrypted
+   * @param secret                 secret used to encrypt the password
+   * @tparam A                     generic parameter for the password
+   * @tparam B                     generic parameter for the secret
+   *  @return                      a string representing the password decrypted
+   */
+  override def decrypt[A, B](encryptedPassword: A, secret:B): String =
+    crypto(EncryptionMode.Decryption, encryptedPassword, secret)
+
+  /**
+   * Method used to perform the either the encryption, or the decryption task
+   *
+   * @param mode                    Mode in with the method must operate, either as a decrypter or an encrypter
+   * @param password                Password to be encrypted or decrypted
+   * @param secret                  Secred to encrypt or decrypt the password
+   * @tparam A                      Generic parameter for the password
+   * @tparam B                      Generic parameter for the secret
+   * @return                        A string representing the password either encrypted or decrypted
+   */
+  def crypto[A, B](mode:EncryptionMode, password: A, secret: B): String
```

Comment:
If not used outside the class, it should be made protected:
```suggestion
  protected def crypto[A, B](mode:EncryptionMode, password: A, secret: B): String
```